### PR TITLE
[API] Cache warmup speedup & refactoring

### PR DESCRIPTION
**Pull request by hyunsooda**
## Proposed changes

API `debug.startWarmUp` loads trie data by starting to read latest state trie, leading fills up trie cache. This PR refactors the implementation and also contributes speedup for cache warmup loop.

The key insight was switching the code below

```go
for i := 0; i < n; i++ {
    ch <- 1
}
```
to
```go
sum := 0
for i := 0; i < n; i++ {
    sum++
}
ch <- sum
```

This modification experimentally makes 1.2x - 1.3x speedup.

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [ ] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [ ] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [ ] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [ ] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- Please leave the issue numbers or links related to this PR here.

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: blockchain/state_migration.go [hunk 1]
```diff
@@ -387,83 +392,96 @@ func (bc *BlockChain) StateMigrationStatus() (bool, uint64, int, int, int, float
...
 	}
-	return nil
 }
 
 // warmUpChecker receives errors from each warm-up goroutine.
 // If it receives a nil error, it means a child goroutine is successfully terminated.
 // It also periodically checks and logs warm-up progress.
-func (bc *BlockChain) warmUpChecker(mainTrieDB *statedb.Database, numChildren int,
-	resultCh chan struct{}, errCh chan error,
+func (bc *BlockChain) wramUpTrieCache(mainTrieDB *statedb.Database, minLoad uint, nChildren int,
```
Comment (blukat29) [inline blockchain/state_migration.go:430]: *warmUpTrieCache

File: blockchain/state_migration.go [hunk 2]
```diff
@@ -37,6 +37,11 @@ import (
 	"github.com/klaytn/klaytn/storage/statedb"
 )
 
+const (
+	DefaultWarmUpMinLoad = 90
+	DefaultIterateItv    = time.Second * 10
```
Comment (blukat29) [inline blockchain/state_migration.go:42]:
```suggestion
	WarmUpCountInterval    = time.Second * 10
```

**Review by blukat29 — COMMENTED** [review]
LGTM except minor comments.

File: blockchain/state_migration.go (hunk 1 above)
Comment (hyunsooda) [reply blockchain/state_migration.go]: Updated. thanks

File: blockchain/state_migration.go (hunk 2 above)
Comment (hyunsooda) [reply blockchain/state_migration.go]: Let me replace it with `DefaultWarmUpReportInterval`. If it looks not clear, let me know yours again.  thanks.

**Review by blukat29 — APPROVED** [review]

File: blockchain/state_migration.go [hunk 3]
```diff
@@ -494,14 +512,19 @@ func (bc *BlockChain) StartWarmUp() error {
...
 	// run goroutine for each child node
-	resultCh := make(chan struct{}, 10000)
+	resultCh := make(chan int, len(children))
 	errCh := make(chan error)
 	bc.quitWarmUp = make(chan struct{})
 	for _, child := range children {
-		go bc.iterateStateTrie(child, db, resultCh, errCh)
+		stateDB, err := state.New(child.Unextend(), db, nil, nil)
+		if err != nil {
+			return err
```
Comment (ian0371) [inline blockchain/state_migration.go:521]: Unless you're trying to change the error handling, I think it should be:

```suggestion
			continue
```

Comment (hyunsooda) [reply blockchain/state_migration.go]: Thank you. Updated.

**Review by blukat29 — APPROVED** [review]

File: blockchain/state_migration.go [hunk 4]
```diff
@@ -387,83 +392,96 @@ func (bc *BlockChain) StateMigrationStatus() (bool, uint64, int, int, int, float
 	return bc.db.InMigration(), bc.db.MigrationBlockNumber(), bc.readCnt, bc.committedCnt, bc.pendingCnt, bc.progress, bc.migrationErr
 }
 
-// iterateStateTrie runs state.Iterator, generated from the given state trie node hash,
+// iterateStateTrie runs state.Iterator, generated from the given state or storage trie node hash,
```
Comment (hyeonLewis) [inline blockchain/state_migration.go:395]:
```suggestion
// trieWarmUp runs state.Iterator, generated from the given state or storage trie node hash,
```

File: blockchain/state_migration.go [hunk 5]
```diff
@@ -387,83 +392,96 @@ func (bc *BlockChain) StateMigrationStatus() (bool, uint64, int, int, int, float
...
+			resultCh <- nReads
+			nReads = 1
 		default:
+			nReads++
 		}
 	}
-	return nil
 }
 
 // warmUpChecker receives errors from each warm-up goroutine.
```
Comment (hyeonLewis) [inline blockchain/state_migration.go:427]:
```suggestion
// warmUpTrieCache receives errors from each warm-up goroutine.
```

**Review by hyeonLewis — APPROVED** [review]

**Comment by hyunsooda** [comment]
Found a syntax error by the update of https://github.com/klaytn/klaytn/pull/2001

**Review by blukat29 — APPROVED** [review]

**Review by hyeonLewis — APPROVED** [review]

**Pull request merged**